Input Component & Select Component Update

andrew-frueh opened the pull request:
Resolves #14. This was originally meant to only be a PR for a new Input text component, but as I got into it I realized that a significant number of our forms use both input text and select boxes.  So it's important our select component and our input components look and act very similarly.  Also, the select component was one of the first added to Cashmere, so it was missing some of our newer practices (ViewEncapsulation to none, not setting style elements directly, etc).

The markup for the Select component was a little odd too in hindsight.  An array of options was being passed to the component in the tag.  But that's pretty limiting because a dev can't set the select option values themselves, they were automatically being set to the array name.  So I've adjusted the markup to let the user set the `<option>` tags themselves.  And to make the input component, I've added the ability to add a label.

Last, the trickiest part of the input component was deciding how to do validation and warning messages.  HTML has a lot of ways to validate text fields directly, but I didn't really want to lock the component into any of them - because different apps will likely want to validate for different things.  It seems that most of our apps validate after a submit click, so I included a `valid` parameter on both the input and select components.  If set to false, it highlights the component and displays any appropriate warning messages.

corytak commented on docs/app/components/input/input-demo.component.html:
```diff
@@ -0,0 +1,53 @@
+        <h5>Example</h5>
+        <h4>Input with Validation</h4>
+        <form (ngSubmit)="onSubmit()">
+            <hc-input [valid]="validCheck" required="true">
```
Since `hc-input` is tied into angular forms, we probably don't need to add our own valid input as it's given to us for free. https://angular.io/guide/form-validation

andrew-frueh replied:
Good suggestion Cory - just recapping what we discussed on slack here.  Instead of hc-input being a separate component, I'll adjust so that it actually extends the base input component to `input [hc-input]` - to take advantage of all the built in form validation.  Then in addition I'll create `label [hc-label]` - which is better because it can be used by both select and input that way.  And probably keep hcErrors as a directive that can be used by any form component.  I'll also take a look and see if it would be possible to be consistent with our select component and switch it to `select [hc-select]`.

corytak commented on docs/app/components/select/select-demo.component.ts:
```diff
@@ -3,15 +3,12 @@ import { Component, ViewEncapsulation } from '@angular/core';
+
 export class SelectDemoComponent {
-    showTemplate: boolean = true;
+    validCheck: boolean = true;
```
I would remove `validCheck` and `toggleValidate()` and rely on angular forms functionality instead.

corytak commented on lib/src/input/input.component.ts:
```diff
@@ -0,0 +1,25 @@
+
+    set valid(validVal) { this._valid = anyToBoolean(validVal); }
+
+    @Input() get required(): boolean { return this._required; }
```
Just like with `valid`, `required` is a Validator available by default from angular forms. it might be nice to hook into that native functionality.

corytak commented on lib/src/select/select.component.ts:
```diff
@@ -13,43 +15,45 @@ import { ControlValueAccessor, NG_VALUE_ACCESSOR } from '@angular/forms';
+
+    set required(isRequired) { this._required = anyToBoolean(isRequired); }
+
+    set disabled(isDisabled) { this._disabled = anyToBoolean(isDisabled); }
```
You may also want to look into reactive/template based forms for setting disabled.

andrew-frueh commented:
@corytak - I've updated the `hc-input` component to be an attribute selector on `<input>` as you suggested, and did the same with `hc-label` on the `<label>` element.  I didn't add another vertical tab for label because it's more of a utility component.  But I did call it out in the documentation on both the Input and Select pages and put it into the example code.

I kept a `required` property on `hc-label` because as best as I can tell, the native label element doesn't have a required value.  So that flag is there to put a red asterisk on the label if set to true.

Also, I kept a `highlight` property on the label so that the dev can decide when they want to put a red highlight around the element.  I could set it to automatically come on with `:invalid` on the css, but it seems like different instances might require validation at different times, so I figured this would be more flexible.  But let me know if I'm overthinking that.

Also updated the Select component's documentation to be markdown on this push as well.

corykon commented on docs/app/components/select/select-demo.component.html:
```diff
@@ -15,68 +15,41 @@ <h5>About</h5>
-                    [options]="['Philadelphia', 'Atlanta', 'Salt Lake City', 'Chicago', 'Orlando']"></hc-select>
+            <label hc-label>Facility Location:</label>
+            <hc-select placeholder="Select a city:">
+                <option>Philadelphia</option>
```
This is a great API change 👍

corykon commented on lib/src/input/input.component.scss:
```diff
@@ -0,0 +1,30 @@
+@import '../sass/mixins.scss';
+
+.hc-input {
+    color: $text;
```
Future PR maybe, just noting it while it's on my mind... would be sweet if there were a "form-controls" sass file containing things that are common across controls... like the styles for borders (default vs error state), font size, padding, etc. This could be a set of variables that each component pulls in, or perhaps standalone sass file that people could use even if they for some reason can't use the angular components but won't style guide approved form controls.

andrew-frueh replied:
That's a great idea.  I had to add the error messages css to the typography css file, but it didn't really seem like it belonged there.  So a forms.scss would be really useful.  If you remember, can you add that as an issue once this PR gets merged?
